### ISSUE TITLE
Puma 5.0 fails to start

### ISSUE THREAD

**Issue by kzkn** [ajgon/opsworks_ruby#244]
### Expected Behavior

Succeeds to start.

### Actual Behavior

It fails to start. When then, it got the following stacktrace:

```
$ ./puma.service start
Set Puma process UID to 5000
Starting Puma rails
(snip)
cd /srv/www/rails/current && /usr/local/bin/bundle exec puma -C /srv/www/rails/shared/config/puma.rb
Traceback (most recent call last):
    12: from /srv/www/rails/shared/vendor/bundle/ruby/2.6.0/bin/puma:23:in `<main>'
(snip)
     1: from /srv/www/rails/shared/vendor/bundle/ruby/2.6.0/gems/puma-5.0.0/lib/puma/dsl.rb:47:in `instance_eval'
/srv/www/rails/shared/config/puma.rb:18:in `_load_from': undefined method `daemonize' for #<Puma::DSL:0x00005573e307e2a0> (NoMethodError)
```

Puma 5.0 has remove the daemonization. This is the cause of the above error.
https://github.com/puma/puma/blob/master/5.0-Upgrade.md#upgrade

### Custom JSON
```json
"appserver": {
    "dot_env": true
},
```

**Comment by kzkn** [comment]
bump

**Comment by ajgon** [comment]
@kzkn This issue has been fixed in most recent (v1.20.0) release. Please let me know if you have any problems.

**Comment by willkoehler** [comment]
@ajgon this may be an edge case, and I'm not really sure the exact conditions. But after upgrading `opsworks_ruby` to a version with this change, I found that, when I deployed, the puma app servers did not restart. I had to restart manually after each deploy with:

```
sudo su deploy
cd /srv/www/yourapp/current
bundle exec pumactl -P /srv/www/yourapp/shared/pids/puma.pid restart
```

I'm guessing that's because my existing AWS instance is still running Puma as a background process and the new `opsworks_ruby` couldn't restart that process because of this change. As a test, I'm rebuilding a new instance from scratch on the latest version of `opsworks_ruby`. I'll update when that's completed.

So the edge case would be upgrading `opsworks_ruby` past this PR on an existing, running, instance. Probably not worth fixing. But might be worth a warning somewhere?

**Comment by willkoehler** [comment]
Continuing in another comment because that one's getting long 😂

Syslog has interesting stuff: See `'puma_shayhi' failed to start (exit status -1)`

```
Jan 31 21:21:21 rails1 monit[929]: 'puma_shayhi' restart on user request
Jan 31 21:21:21 rails1 monit[929]: Monit daemon with PID 929 awakened
Jan 31 21:21:21 rails1 monit[929]: Awakened by User defined signal 1
Jan 31 21:21:21 rails1 monit[929]: 'puma_shayhi' trying to restart
Jan 31 21:21:21 rails1 monit[929]: 'puma_shayhi' stop: '/bin/sh -c cat /run/lock/shayhi/puma.pid | xargs --no-run-if-empty kill -QUIT; sleep 5'
Jan 31 21:21:21 rails1 monit: Skipping 'allow localhost' -- host resolved to [::ffff:127.0.0.1] which is present in ACL already
Jan 31 21:21:21 rails1 monit[929]: 'delayed_job_shayhi-1' restart on user request
Jan 31 21:21:21 rails1 monit[929]: Monit daemon with PID 929 awakened
Jan 31 21:21:23 rails1 systemd[1]: Stopping nginx - high performance web server...
Jan 31 21:21:23 rails1 systemd[1]: Stopped nginx - high performance web server.
Jan 31 21:21:23 rails1 systemd[1]: Starting nginx - high performance web server...
Jan 31 21:21:23 rails1 nginx[8212]: nginx: [warn] the "ssl" directive is deprecated, use the "listen ... ssl" directive instead in /etc/nginx/sites-enabled/shayhi.conf:74
Jan 31 21:21:23 rails1 systemd[1]: Started nginx - high performance web server.
Jan 31 21:21:24 rails1 systemd[1]: Reloading nginx - high performance web server.
Jan 31 21:21:24 rails1 systemd[1]: Reloaded nginx - high performance web server.
Jan 31 21:21:26 rails1 monit[929]: 'puma_shayhi' start: '/bin/sh -c cd /srv/www/shayhi/current && MAILER_URL_HOST="staging.shayhi.com" RAILS_ENV="production" HOME="/home/deploy" USER="deploy" bundle exec puma -C /srv/www/shayhi/shared/config/puma.rb | logger -t puma-shayhi'
Jan 31 21:21:29 rails1 puma-shayhi: [8266] Puma starting in cluster mode...
Jan 31 21:21:29 rails1 puma-shayhi: [8266] * Puma version: 5.2.0 (ruby 2.7.2-p137) ("Fettisdagsbulle")
Jan 31 21:21:29 rails1 puma-shayhi: [8266] *  Min threads: 0
Jan 31 21:21:29 rails1 puma-shayhi: [8266] *  Max threads: 16
Jan 31 21:21:29 rails1 puma-shayhi: [8266] *  Environment: production
Jan 31 21:21:29 rails1 puma-shayhi: [8266] *   Master PID: 8266
Jan 31 21:21:29 rails1 puma-shayhi: [8266] *      Workers: 4
Jan 31 21:21:29 rails1 puma-shayhi: [8266] *     Restarts: (✔) hot (✖) phased
Jan 31 21:21:29 rails1 puma-shayhi: [8266] * Preloading application
Jan 31 21:21:30 rails1 puma-shayhi: [8266] * Listening on unix:///srv/www/shayhi/shared/sockets/puma.sock
Jan 31 21:21:30 rails1 puma-shayhi: [8266] ! WARNING: Detected 2 Thread(s) started in app boot:
Jan 31 21:21:30 rails1 puma-shayhi: [8266] ! #<Thread:0x000055ac8652a8b8 /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/activerecord-6.0.3.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:334 sleep> - /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/activerecord-6.0.3.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:337:in `sleep'
Jan 31 21:21:30 rails1 puma-shayhi: [8266] ! #<Rack::MiniProfiler::FileStore::CacheCleanupThread:0x000055ac86dba960 /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/rack-mini-profiler-1.1.5/lib/mini_profiler/storage/file_store.rb:63 sleep> - /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/rack-mini-profiler-1.1.5/lib/mini_profiler/storage/file_store.rb:81:in `sleep'
Jan 31 21:21:30 rails1 puma-shayhi: [8266] Use Ctrl-C to stop
Jan 31 21:22:01 rails1 CRON[8329]: (root) CMD (flock --nonblock /var/lib/aws/opsworks/lockrun.lock /opt/aws/opsworks/current/bin/opsworks-agent-updater)
Jan 31 21:22:01 rails1 CRON[8330]: (root) CMD (/var/awslogs/bin/awslogs-nanny.sh > /dev/null 2>&1)
Jan 31 21:22:56 rails1 monit[929]: 'puma_shayhi' failed to start (exit status -1) -- Program '/bin/sh -c cd /srv/www/shayhi/current && MAILER_URL_HOST="staging.shayhi.com" RAILS_ENV="production" HOME="/home/deploy" USER="deploy" bundle exec puma -C /srv/www/shayhi/shared/config/puma.rb | logger -t puma-shayhi' timed out after 1.500 m
Jan 31 21:22:56 rails1 monit[929]: 'puma_shayhi' restart action failed
Jan 31 21:22:56 rails1 monit[929]: 'delayed_job_shayhi-1' trying to restart
Jan 31 21:22:56 rails1 monit[929]: 'delayed_job_shayhi-1' stop: '/bin/su - deploy -c cd /srv/www/shayhi/current && MAILER_URL_HOST="staging.shayhi.com" RAILS_ENV="production" HOME="/home/deploy" USER="deploy" bin/delayed_job stop --pid-dir=/run/lock/shayhi/ -i 0'
Jan 31 21:22:56 rails1 systemd[1]: Started Session c5 of user deploy.
```

But some deploys it doesn't even look like puma tries to restart 🤔 Waaa?

```
Jan 31 22:20:01 rails1 CRON[12522]: (root) CMD (/var/awslogs/bin/awslogs-nanny.sh > /dev/null 2>&1)
Jan 31 22:20:01 rails1 CRON[12523]: (root) CMD (flock --nonblock /var/lib/aws/opsworks/lockrun.lock /opt/aws/opsworks/current/bin/opsworks-agent-updater)
Jan 31 22:20:26 rails1 dbus-daemon[875]: [system] Activating via systemd: service name='org.freedesktop.hostname1' unit='dbus-org.freedesktop.hostname1.service' requested by ':1.155' (uid=0 pid=13058 comm="/usr/bin/hostnamectl " label="unconfined")
Jan 31 22:20:26 rails1 systemd[1]: Starting Hostname Service...
Jan 31 22:20:26 rails1 dbus-daemon[875]: [system] Successfully activated service 'org.freedesktop.hostname1'
Jan 31 22:20:26 rails1 systemd[1]: Started Hostname Service.
Jan 31 22:20:29 rails1 monit: Skipping 'allow localhost' -- host resolved to [::ffff:127.0.0.1] which is present in ACL already
Jan 31 22:20:29 rails1 monit[13094]: Reinitializing monit daemon
Jan 31 22:20:29 rails1 monit[929]: Reinitializing Monit -- control file '/etc/monit/monitrc'
Jan 31 22:20:29 rails1 monit: Skipping 'allow localhost' -- host resolved to [::ffff:127.0.0.1] which is present in ACL already
Jan 31 22:20:29 rails1 monit[13097]: Reinitializing monit daemon
Jan 31 22:20:29 rails1 monit[929]: Skipping 'allow localhost' -- host resolved to [::ffff:127.0.0.1] which is present in ACL already
Jan 31 22:20:29 rails1 monit[929]: 'rails1' Monit reloaded
Jan 31 22:21:01 rails1 CRON[13238]: (root) CMD (/var/awslogs/bin/awslogs-nanny.sh > /dev/null 2>&1)
Jan 31 22:21:01 rails1 CRON[13240]: (root) CMD (flock --nonblock /var/lib/aws/opsworks/lockrun.lock /opt/aws/opsworks/current/bin/opsworks-agent-updater)
Jan 31 22:21:27 rails1 monit: Skipping 'allow localhost' -- host resolved to [::ffff:127.0.0.1] which is present in ACL already
Jan 31 22:21:27 rails1 monit[929]: 'puma_shayhi' restart on user request
Jan 31 22:21:27 rails1 monit[929]: Monit daemon with PID 929 awakened
Jan 31 22:21:27 rails1 monit[929]: Awakened by User defined signal 1
Jan 31 22:21:27 rails1 monit[929]: Reinitializing Monit -- control file '/etc/monit/monitrc'
Jan 31 22:21:27 rails1 monit: Skipping 'allow localhost' -- host resolved to [::ffff:127.0.0.1] which is present in ACL already
Jan 31 22:21:27 rails1 monit[929]: 'delayed_job_shayhi-1' restart on user request
Jan 31 22:21:27 rails1 monit[929]: Monit daemon with PID 929 awakened
Jan 31 22:21:27 rails1 monit[929]: Skipping 'allow localhost' -- host resolved to [::ffff:127.0.0.1] which is present in ACL already
Jan 31 22:21:27 rails1 monit[929]: 'rails1' Monit reloaded
Jan 31 22:21:29 rails1 systemd[1]: Stopping nginx - high performance web server...
Jan 31 22:21:29 rails1 systemd[1]: Stopped nginx - high performance web server.
Jan 31 22:21:29 rails1 systemd[1]: Starting nginx - high performance web server...
Jan 31 22:21:29 rails1 nginx[13349]: nginx: [warn] the "ssl" directive is deprecated, use the "listen ... ssl" directive instead in /etc/nginx/sites-enabled/shayhi.conf:74
Jan 31 22:21:29 rails1 systemd[1]: Started nginx - high performance web server.
Jan 31 22:21:29 rails1 systemd[1]: Reloading nginx - high performance web server.
Jan 31 22:21:29 rails1 systemd[1]: Reloaded nginx - high performance web server.
```

**Comment by willkoehler** [comment]
Interesting. If I just try to run the monit command manually I get:

```
deploy@rails1:/srv/www/shayhi/current$ MAILER_URL_HOST="staging.shayhi.com" RAILS_ENV="production" HOME="/home/deploy" USER="deploy" bundle exec puma -C /srv/www/shayhi/shared/config/puma.rb
[13572] Puma starting in cluster mode...
[13572] * Puma version: 5.2.0 (ruby 2.7.2-p137) ("Fettisdagsbulle")
[13572] *  Min threads: 0
[13572] *  Max threads: 16
[13572] *  Environment: production
[13572] *   Master PID: 13572
[13572] *      Workers: 4
[13572] *     Restarts: (✔) hot (✖) phased
[13572] * Preloading application
bundler: failed to load command: puma (/srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/bin/puma)
Traceback (most recent call last):
	25: from /usr/lib/fullstaq-ruby/versions/2.7/bin/bundle:23:in `<main>'
	24: from /usr/lib/fullstaq-ruby/versions/2.7/bin/bundle:23:in `load'
	23: from /usr/lib/fullstaq-ruby/versions/2.7/lib/ruby/gems/2.7.0/gems/bundler-2.2.7/exe/bundle:37:in `<top (required)>'
	22: from /usr/lib/fullstaq-ruby/versions/2.7/lib/ruby/gems/2.7.0/gems/bundler-2.2.7/lib/bundler/friendly_errors.rb:130:in `with_friendly_errors'
	21: from /usr/lib/fullstaq-ruby/versions/2.7/lib/ruby/gems/2.7.0/gems/bundler-2.2.7/exe/bundle:49:in `block in <top (required)>'
	20: from /usr/lib/fullstaq-ruby/versions/2.7/lib/ruby/gems/2.7.0/gems/bundler-2.2.7/lib/bundler/cli.rb:24:in `start'
	19: from /usr/lib/fullstaq-ruby/versions/2.7/lib/ruby/gems/2.7.0/gems/bundler-2.2.7/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
	18: from /usr/lib/fullstaq-ruby/versions/2.7/lib/ruby/gems/2.7.0/gems/bundler-2.2.7/lib/bundler/cli.rb:30:in `dispatch'
	17: from /usr/lib/fullstaq-ruby/versions/2.7/lib/ruby/gems/2.7.0/gems/bundler-2.2.7/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
	16: from /usr/lib/fullstaq-ruby/versions/2.7/lib/ruby/gems/2.7.0/gems/bundler-2.2.7/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
	15: from /usr/lib/fullstaq-ruby/versions/2.7/lib/ruby/gems/2.7.0/gems/bundler-2.2.7/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
	14: from /usr/lib/fullstaq-ruby/versions/2.7/lib/ruby/gems/2.7.0/gems/bundler-2.2.7/lib/bundler/cli.rb:494:in `exec'
	13: from /usr/lib/fullstaq-ruby/versions/2.7/lib/ruby/gems/2.7.0/gems/bundler-2.2.7/lib/bundler/cli/exec.rb:28:in `run'
	12: from /usr/lib/fullstaq-ruby/versions/2.7/lib/ruby/gems/2.7.0/gems/bundler-2.2.7/lib/bundler/cli/exec.rb:63:in `kernel_load'
	11: from /usr/lib/fullstaq-ruby/versions/2.7/lib/ruby/gems/2.7.0/gems/bundler-2.2.7/lib/bundler/cli/exec.rb:63:in `load'
	10: from /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/bin/puma:23:in `<top (required)>'
	 9: from /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/bin/puma:23:in `load'
	 8: from /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/puma-5.2.0/bin/puma:10:in `<top (required)>'
	 7: from /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/puma-5.2.0/lib/puma/cli.rb:80:in `run'
	 6: from /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/puma-5.2.0/lib/puma/launcher.rb:181:in `run'
	 5: from /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/puma-5.2.0/lib/puma/cluster.rb:342:in `run'
	 4: from /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/puma-5.2.0/lib/puma/runner.rb:144:in `load_and_bind'
	 3: from /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/puma-5.2.0/lib/puma/binder.rb:152:in `parse'
	 2: from /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/puma-5.2.0/lib/puma/binder.rb:152:in `each'
	 1: from /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/puma-5.2.0/lib/puma/binder.rb:208:in `block in parse'
/srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/puma-5.2.0/lib/puma/binder.rb:369:in `add_unix_listener': There is already a server bound to: /srv/www/shayhi/shared/sockets/puma.sock (RuntimeError)
```

**Comment by willkoehler** [comment]
I can restart manually on the command-line using the same commands that are in monit

```
deploy@rails1:/srv/www/shayhi/current$ MAILER_URL_HOST="staging.shayhi.com" RAILS_ENV="production" HOME="/home/deploy" USER="deploy" bundle exec puma -C /srv/www/shayhi/shared/config/puma.rb
[13801] Puma starting in cluster mode...
[13801] * Puma version: 5.2.0 (ruby 2.7.2-p137) ("Fettisdagsbulle")
[13801] *  Min threads: 0
[13801] *  Max threads: 16
[13801] *  Environment: production
[13801] *   Master PID: 13801
[13801] *      Workers: 4
[13801] *     Restarts: (✔) hot (✖) phased
[13801] * Preloading application
[13801] * Listening on unix:///srv/www/shayhi/shared/sockets/puma.sock
[13801] ! WARNING: Detected 2 Thread(s) started in app boot:
[13801] ! #<Thread:0x000055bcddb40010 /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/activerecord-6.0.3.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:334 sleep> - /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/activerecord-6.0.3.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:337:in `sleep'
[13801] ! #<Rack::MiniProfiler::FileStore::CacheCleanupThread:0x000055bcdd397c78 /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/rack-mini-profiler-1.1.5/lib/mini_profiler/storage/file_store.rb:63 sleep> - /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/rack-mini-profiler-1.1.5/lib/mini_profiler/storage/file_store.rb:81:in `sleep'
[13801] Use Ctrl-C to stop
```

**Comment by willkoehler** [comment]
But on the next deploy, monit does not restart puma 🤔 

```
Jan 31 22:38:57 rails1 monit[929]: 'puma_shayhi' process is running with pid 14032
Jan 31 22:38:57 rails1 monit[929]: 'puma_shayhi' process is running after previous exec error (slow starting or manually recovered?)
Jan 31 22:39:01 rails1 CRON[14157]: (root) CMD (flock --nonblock /var/lib/aws/opsworks/lockrun.lock /opt/aws/opsworks/current/bin/opsworks-agent-updater)
Jan 31 22:39:01 rails1 CRON[14158]: (root) CMD (/var/awslogs/bin/awslogs-nanny.sh > /dev/null 2>&1)
Jan 31 22:40:01 rails1 CRON[14185]: (root) CMD (flock --nonblock /var/lib/aws/opsworks/lockrun.lock /opt/aws/opsworks/current/bin/opsworks-agent-updater)
Jan 31 22:40:01 rails1 CRON[14186]: (root) CMD (/var/awslogs/bin/awslogs-nanny.sh > /dev/null 2>&1)
Jan 31 22:41:01 rails1 CRON[14218]: (root) CMD (flock --nonblock /var/lib/aws/opsworks/lockrun.lock /opt/aws/opsworks/current/bin/opsworks-agent-updater)
Jan 31 22:41:01 rails1 CRON[14219]: (root) CMD (/var/awslogs/bin/awslogs-nanny.sh > /dev/null 2>&1)
Jan 31 22:41:10 rails1 dbus-daemon[875]: [system] Activating via systemd: service name='org.freedesktop.hostname1' unit='dbus-org.freedesktop.hostname1.service' requested by ':1.163' (uid=0 pid=14750 comm="/usr/bin/hostnamectl " label="unconfined")
Jan 31 22:41:10 rails1 systemd[1]: Starting Hostname Service...
Jan 31 22:41:10 rails1 dbus-daemon[875]: [system] Successfully activated service 'org.freedesktop.hostname1'
Jan 31 22:41:10 rails1 systemd[1]: Started Hostname Service.
Jan 31 22:41:12 rails1 monit: Skipping 'allow localhost' -- host resolved to [::ffff:127.0.0.1] which is present in ACL already
Jan 31 22:41:12 rails1 monit[14786]: Reinitializing monit daemon
Jan 31 22:41:12 rails1 monit[929]: Reinitializing Monit -- control file '/etc/monit/monitrc'
Jan 31 22:41:12 rails1 monit: Skipping 'allow localhost' -- host resolved to [::ffff:127.0.0.1] which is present in ACL already
Jan 31 22:41:12 rails1 monit[14789]: Reinitializing monit daemon
Jan 31 22:41:13 rails1 monit[929]: Skipping 'allow localhost' -- host resolved to [::ffff:127.0.0.1] which is present in ACL already
Jan 31 22:41:13 rails1 monit[929]: 'rails1' Monit reloaded
Jan 31 22:42:01 rails1 CRON[14993]: (root) CMD (/var/awslogs/bin/awslogs-nanny.sh > /dev/null 2>&1)
Jan 31 22:42:01 rails1 CRON[14994]: (root) CMD (flock --nonblock /var/lib/aws/opsworks/lockrun.lock /opt/aws/opsworks/current/bin/opsworks-agent-updater)
Jan 31 22:42:12 rails1 monit: Skipping 'allow localhost' -- host resolved to [::ffff:127.0.0.1] which is present in ACL already
Jan 31 22:42:12 rails1 monit[929]: 'puma_shayhi' restart on user request
Jan 31 22:42:12 rails1 monit[929]: Monit daemon with PID 929 awakened
Jan 31 22:42:12 rails1 monit[929]: Awakened by User defined signal 1
Jan 31 22:42:12 rails1 monit[929]: Reinitializing Monit -- control file '/etc/monit/monitrc'
Jan 31 22:42:12 rails1 monit: Skipping 'allow localhost' -- host resolved to [::ffff:127.0.0.1] which is present in ACL already
Jan 31 22:42:12 rails1 monit[929]: 'delayed_job_shayhi-1' restart on user request
Jan 31 22:42:12 rails1 monit[929]: Monit daemon with PID 929 awakened
Jan 31 22:42:12 rails1 monit[929]: Skipping 'allow localhost' -- host resolved to [::ffff:127.0.0.1] which is present in ACL already
Jan 31 22:42:12 rails1 monit[929]: 'rails1' Monit reloaded
Jan 31 22:42:14 rails1 systemd[1]: Stopping nginx - high performance web server...
Jan 31 22:42:14 rails1 systemd[1]: Stopped nginx - high performance web server.
Jan 31 22:42:14 rails1 systemd[1]: Starting nginx - high performance web server...
```

**Comment by willkoehler** [comment]
I can manually restart puma via monit from the command line

```
deploy@rails1:/srv/www/shayhi/current$ sudo monit restart puma_shayhi
Skipping 'allow localhost' -- host resolved to [::ffff:127.0.0.1] which is present in ACL already
deploy@rails1:/srv/www/shayhi/current$
```

Which results in this in syslog
```
Jan 31 22:50:54 rails1 monit: Skipping 'allow localhost' -- host resolved to [::ffff:127.0.0.1] which is present in ACL already
Jan 31 22:51:00 rails1 monit: Skipping 'allow localhost' -- host resolved to [::ffff:127.0.0.1] which is present in ACL already
Jan 31 22:51:00 rails1 monit[929]: 'puma_shayhi' restart on user request
Jan 31 22:51:00 rails1 monit[929]: Monit daemon with PID 929 awakened
Jan 31 22:51:00 rails1 monit[929]: Awakened by User defined signal 1
Jan 31 22:51:00 rails1 monit[929]: 'puma_shayhi' trying to restart
Jan 31 22:51:00 rails1 monit[929]: 'puma_shayhi' stop: '/bin/sh -c cat /run/lock/shayhi/puma.pid | xargs --no-run-if-empty kill -QUIT; sleep 5'
Jan 31 22:51:01 rails1 CRON[15450]: (root) CMD (flock --nonblock /var/lib/aws/opsworks/lockrun.lock /opt/aws/opsworks/current/bin/opsworks-agent-updater)
Jan 31 22:51:01 rails1 CRON[15451]: (root) CMD (/var/awslogs/bin/awslogs-nanny.sh > /dev/null 2>&1)
Jan 31 22:51:05 rails1 monit[929]: 'puma_shayhi' start: '/bin/sh -c cd /srv/www/shayhi/current && MAILER_URL_HOST="staging.shayhi.com" RAILS_ENV="production" HOME="/home/deploy" USER="deploy" bundle exec puma -C /srv/www/shayhi/shared/config/puma.rb | logger -t puma-shayhi'
Jan 31 22:51:08 rails1 puma-shayhi: [15460] Puma starting in cluster mode...
Jan 31 22:51:08 rails1 puma-shayhi: [15460] * Puma version: 5.2.0 (ruby 2.7.2-p137) ("Fettisdagsbulle")
Jan 31 22:51:08 rails1 puma-shayhi: [15460] *  Min threads: 0
Jan 31 22:51:08 rails1 puma-shayhi: [15460] *  Max threads: 16
Jan 31 22:51:08 rails1 puma-shayhi: [15460] *  Environment: production
Jan 31 22:51:08 rails1 puma-shayhi: [15460] *   Master PID: 15460
Jan 31 22:51:08 rails1 puma-shayhi: [15460] *      Workers: 4
Jan 31 22:51:08 rails1 puma-shayhi: [15460] *     Restarts: (✔) hot (✖) phased
Jan 31 22:51:08 rails1 puma-shayhi: [15460] * Preloading application
Jan 31 22:51:09 rails1 puma-shayhi: [15460] * Listening on unix:///srv/www/shayhi/shared/sockets/puma.sock
Jan 31 22:51:09 rails1 puma-shayhi: [15460] ! WARNING: Detected 2 Thread(s) started in app boot:
Jan 31 22:51:09 rails1 puma-shayhi: [15460] ! #<Thread:0x00005607871a9d68 /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/activerecord-6.0.3.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:334 sleep> - /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/activerecord-6.0.3.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:337:in `sleep'
Jan 31 22:51:09 rails1 puma-shayhi: [15460] ! #<Rack::MiniProfiler::FileStore::CacheCleanupThread:0x0000560787952c40 /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/rack-mini-profiler-1.1.5/lib/mini_profiler/storage/file_store.rb:63 sleep> - /srv/www/shayhi/shared/vendor/bundle/ruby/2.7.0/gems/rack-mini-profiler-1.1.5/lib/mini_profiler/storage/file_store.rb:81:in `sleep'
Jan 31 22:51:09 rails1 puma-shayhi: [15460] Use Ctrl-C to stop
```

**Comment by willkoehler** [comment]
My best guess at this point is there is a race condition between restarting puma and restarting monit as evidence with the `Reinitializing Monit -- control file '/etc/monit/monitrc'` log entry immediately after requesting the restart vs `'puma_shayhi' trying to restart` when monit is not being restarted. See previous comment with difference between the two logs.

```
Feb  1 01:42:27 rails1 monit[929]: 'puma_shayhi' restart on user request
Feb  1 01:42:27 rails1 monit[929]: Monit daemon with PID 929 awakened
Feb  1 01:42:27 rails1 monit[929]: Awakened by User defined signal 1
Feb  1 01:42:27 rails1 monit[929]: Reinitializing Monit -- control file '/etc/monit/monitrc'
Feb  1 01:42:27 rails1 monit: Skipping 'allow localhost' -- host resolved to [::ffff:127.0.0.1] which is present in ACL already
```

**Comment by willkoehler** [comment]
Oh, It looks like @kzkn is also on top of this, with a proposed fix in #251

**Comment by ajgon** [comment]
Wow... this is clearly the best bug report this project has seen :bowing_man: Glad that you and @kzkn figured this out - I've merged his fix and will release it soon :)
